validator/src/validator.rs

```diff
@@ -50,6 +49,51 @@ pub fn run_validator(config: Config) {
         .expect("Invalid Genesis Hash");
 
     loop {
+        // Check if there is a pending block to sign
+        let chain_account: ChainData = program.account(chain).unwrap();
+        if chain_account.pending_block().unwrap().is_some() {
```

Comment:
As it stands, we’ll keep trying to sign this block over and over until it gets finalised. We should at least remember height of the last signed block and compare it with pending.

Reply:
it checks for pending block and the signature. If it contains the signature it doesnt sign the block again and then moves on to listen to `new_block` events.

Reply:
but yes, we need to keep record of last signed block.